src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs

```diff
@@ -403,6 +403,38 @@ private static bool IsTrimmableAssembly(ModuleDesc assembly)
             return false;
         }
 
+        public override void GetDependenciesDueToEETypePresence(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
+        {
+            base.GetDependenciesDueToEETypePresence(ref dependencies, factory, type);
+
+            try
+            {
+                if (type.HasBaseType)
+                {
+                    GetDataFlowDependenciesForInstantiation(ref dependencies, factory, type.BaseType, type);
+                }
+
+                foreach (var runtimeInterface in type.RuntimeInterfaces)
+                {
+                    GetDataFlowDependenciesForInstantiation(ref dependencies, factory, runtimeInterface, type);
+                }
+            }
+            catch (TypeSystemException)
+            {
+                // Wasn't able to do dataflow because of missing references or something like that.
+                // This likely won't compile either, so we don't care about missing dependencies.
+            }
+        }
+
+        private void GetDataFlowDependenciesForInstantiation(ref DependencyList dependencies, NodeFactory factory, TypeDesc type, TypeDesc contextType)
+        {
+            if (type.HasInstantiation && FlowAnnotations.HasGenericParameterAnnotation(type))
+            {
+                TypeDesc instantiatedType = type.InstantiateSignature(contextType.Instantiation, Instantiation.Empty);
```

Comment:
I'll look into this, but it's probably a left over before I decided to use tokens in the tests. Without this sometimes the type was reported as
```
SomeGeneric`1
```
While what we want is something like
```
SomeGeneric<T>
```

Reply:
This is actually necessary - it turns `BaseType<!0>` into `BaseType<TOuter>`. Otherwise we would have to map the signature variables to the derived type's generic parameters manually.

But you comment brought up a good point, that this logic was actually running potentially multiple times per type (for each instantiation) which could lead to duplicate or misleading warnings. So I rewrote this into a new dependency node and only run the analysis on the type definition (instantiations are just confusing, because they would have the base type also instantiated which could lead to misleading warnings).

From what I saw it's a rule that for each instantiated EEType there will be an EEType for the definition as well, but it felt relatively fragile to depend on it (I could not find it in the code written as a hard rule) - so I added a new dependency node which is only create for the type definition (and only if the type is interesting in some way).